tests/tests/preconfirmations.rs

```diff
@@ -0,0 +1,416 @@
+        TransactionStatus::Failure { .. }
+    ));
+}
+
+#[tokio::test]
+async fn preconfirmation__received_after_squeezed_out() {
+    let mut config = Config::local_node();
+    config.block_production = Trigger::Never;
+    let mut rng = rand::thread_rng();
+
+    // Given
+    // Disable UTXO validation in TxPool so that the transaction is squeezed out by
+    // block production
+    config.debug = true;
+    config.txpool.utxo_validation = false;
+    config.utxo_validation = true;
+    let srv = FuelService::new_node(config).await.unwrap();
+    let client = FuelClient::from(srv.bound_address);
+
+    let gas_limit = 1_000_000;
+    let tx = TransactionBuilder::script(
+        vec![op::ret(RegId::ONE)].into_iter().collect(),
+        vec![],
+    )
+    .script_gas_limit(gas_limit)
+    .add_unsigned_coin_input(
+        SecretKey::random(&mut rng),
+        rng.gen(),
+        10,
+        Default::default(),
+        Default::default(),
+    )
+    .add_output(Output::Change {
+        to: Default::default(),
+        amount: 0,
+        asset_id: Default::default(),
+    })
+    .finalize_as_transaction();
+
+    let tx_id = tx.id(&Default::default());
+    let mut tx_statuses_subscriber =
+        client.subscribe_transaction_status(&tx_id).await.unwrap();
+    client.submit(&tx).await.unwrap();
+
+    // When
+    assert!(matches!(
+        tx_statuses_subscriber.next().await.unwrap().unwrap(),
+        TransactionStatus::Submitted { .. }
+    ));
+    client.produce_blocks(1, None).await.unwrap();
+    if let TransactionStatus::PreconfirmationSqueezedOut {
+        transaction_id,
+        reason: _,
+    } = tx_statuses_subscriber.next().await.unwrap().unwrap()
+    {
+        // Then
+        assert_eq!(transaction_id, tx_id);
+    } else {
+        panic!("Expected preconfirmation status");
+    }
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+async fn preconfirmation__received_tx_inserted_end_block_open_period() {
+    let mut config = Config::local_node();
+    let block_production_period = Duration::from_secs(1);
+    let address = Address::new([0; 32]);
+
+    config.block_production = Trigger::Open {
+        period: block_production_period,
+    };
+    let srv = FuelService::new_node(config).await.unwrap();
+    let client = FuelClient::from(srv.bound_address);
+
+    // Given
+    let tx = TransactionBuilder::script(
+        vec![op::ret(RegId::ONE)].into_iter().collect(),
+        vec![],
+    )
+    .script_gas_limit(1_000_000)
+    .add_fee_input()
+    .add_output(Output::variable(address, 0, AssetId::default()))
+    .finalize_as_transaction();
+
+    let tx_id = tx.id(&Default::default());
+    let jh = tokio::spawn({
+        let client = client.clone();
+        async move {
+            let _ = client
+                .subscribe_transaction_status(&tx_id)
+                .await
+                .unwrap()
+                .enumerate()
+                .take(3)
```

Comment:
I was wondering if we could avoid hardcoding 3 and try to read all events, just to verify that no more unexpected events are sent.

Reply:
Removed it.